### DISCUSSION TITLE
Using apple accelerate/intel optimized BLAS  with gensim and other numpy numeric computation

### DISCUSSION THREAD

**Discussion by danieltomasz** [DrewThomasson/ebook2audiobook#147] [Ideas]
I wonder how much of the the workload here could be make faster by using platform optimized  BLAS?
I am on MacOS M4Pro and simple ebook (in French)  took me more than 1h to create an audiobook, I wonder if this might by made faster by using platform optimised conda  BLAS installs instead of default one (which might not be platform optimised, depending from platform)

**Comment by ROBERT-MCDOWELL** [comment]
many factors are involved to speed up a process, including GPU drivers, the language itself (python can be slow), the library and components used reacting differently on each OS and hardware.... it's not one recipe for all but many recipe for each....
btw BLAS cannot do anything without optimization from the TTS code itself

**Comment by danieltomasz** [comment]
’m asking because I come from a numerical background, where it makes a significant difference if I use an optimized library or not—sometimes as much as a 10x improvement when using NumPy—without changing much of the code. For example, with Gensim, using an optimized BLAS instead of the default libraries installed via Conda can make Word2Vec faster... but if  the bottleneck here is the TTS code I dont if it will help ... 

Currently numpy installed  in the env installed via script is using default  `openblas64`  (because it is pip installed)
`python -c "import numpy; numpy.show_config()"`

If I will have time I will check if installing some  dependencies directly  from conda forge  and forcing  - libblas=*=*accelerate will help with time

It will probably will be still much slower that GPU, but it still  might cut here  the time (depending what is bottleneck here)

**Comment by ROBERT-MCDOWELL** [comment]
if you have an idea to implement it for Windows/Linux/Mac so please create a PR thanks

**Comment by ROBERT-MCDOWELL** [comment]
btw did you hear about modular? https://www.modular.com/mojo

**Comment by danieltomasz** [comment]
yes, but mojo is rather a separate language, in a big subset interoperable with python; many people  looking for performance now choose instead to rewrite some parts of code in rust and combine it with high level Python (this is probably a thing that I would like to learn in the future)

my idea here is much more modest,  is just to see if I can made the python libraries in the environment  to link to  platform optimized libraries via  explicitly defining it via conda env, so it would   be a really low hanging fruit, and maybe modest improvement for people running code on their CPUI
Numpy 2.0 support it out of the box  via pip as well, but many libraries still uses numpy 1.26

I will  indeed play with this idea and If it help with performance provide a PR

**Comment by ROBERT-MCDOWELL** [comment]
don't you have to recompile all?

**Reply by danieltomasz** [reply]
My experience is when is want to use libraries installed via pip, I indeed need to compile  with numpy <2.0 via 'pip install numpy --no-binary numpy' if I install directly from conda forge I can install optimized libraries within env and numpy/Numba linked to them (MKL in case of Intel machines, Accelerate in case of Apple Silicon et cetera)

**Reply by danieltomasz** [reply]
The install script (line 217) doesn't use conda forge channel to specify required libraries, it just pip install requirements, but using conda provided numpy/numba can have advantages https://numba.readthedocs.io/en/stable/user/performance-tips.html

Not every package is available on conda forge, so the installation process needs the careful checks when mixing conda and pip to avoid problem with dependencies

**Reply by ROBERT-MCDOWELL** [reply]
for now it's not our priority as the script must be the most universal as possible as it must cover all compatible OSes.
we have a lot of more issues to solve and make the project with its essential and crucial functionalities stable.
but you can of course experiment on your side and report the results with a possible patch for specific OS.
